docs/unread_messages.md

```diff
@@ -0,0 +1,50 @@
+# Unread message synchronization
+
+In general displaying unread counts for all streams and topics may require
+downloading an unbounded number of messages. Consider a user who has a muted
+stream or topic and has not read the backlog in a month; to have an accurate
+unread count we would need to load all messages this user has received in the
+past month. This is inefficient for web clients and even more for mobile
+devices.
+
+We work around this by including a list of unread message ids in the initial
+state grouped by relevant conversation keys. This data is included in the
+`unread_msgs` key if both `update_message_flags` and `message` are required
+in the register call.
+
+```
+{
+    "huddles": [
+        {
+            "user_ids_string": "3,4,6",
+            "unread_message_ids": [
+                34
+            ]
+        }
+    ],
+    "streams": [
+        {
+            "stream_id": 1,
+            "topic": "test",
+            "unread_message_ids": [
+                33
+            ]
+        }
+    ],
+    "pms": [
+        {
+            "sender_id": 3,
+            "unread_message_ids": [
+                31,
+                32
+            ]
```

Comment:
I think splitting them up is convenient, since clients may possibly have views that are huddle-only, etc., plus it's a little more convenient to show the slightly different structures for various recipient types if they're each in their own bucket.

We may re-visit this if we get more feedback from the mobile team, but for now I want to keep the structure they already agreed on.